### PR TITLE
Reduction integration

### PR DESCRIPTION
**Pull request by jjsjann123**
reduction integration is now functional, I'm passing the added tests.

I probably digged a big hole in our fake broadcast support in integration. As we are getting better implicit broadcast support (mismatch size/dimension), I should start working on that in next PR.

For now I still need some refactoring on operation registration, currently treating reduction as a special op, which is more of a hack. I'll put a proper enum for operator category (pointwise / broadcast / reduction / pooling?)

### CODE REVIEW

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -331,14 +333,14 @@ def where(x: torch.Tensor, y: torch.Tensor, cond: torch.Tensor):
         where_jit = torch.jit.script(where)
         self._run_helper(where_jit, where, x, y, cond)
 
-        def lerp(x : torch.Tensor, y : torch.Tensor, z : torch.Tensor):
+        def lerp(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
```
Comment (kevinstephano) [inline test/test_jit_cuda_fuser.py:336]: When you specify the types for inputs and outputs in python for torch.script, I thought you also had to specify the output type.  Is that not true?

Comment (jjsjann123) [reply test/test_jit_cuda_fuser.py:336]: You could but I thought that's optional (some of their examples definitely don't have it.)
https://pytorch.org/docs/stable/jit_language_reference.html#variable-resolution

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 2]
```diff
@@ -395,7 +395,7 @@ static TensorView* newForReduction(
 
 TensorView* reductionOp(
     BinaryOpType reduction_op_type,
-    const std::vector<int>& axes,
+    const std::vector<int64_t>& axes,
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: Why are the axes variables being increased from 4 bytes to 8 bytes?  I thought we would only ever have 8+1 positions?  I am guessing there is some mechanical or legality issue with narrowing.

File: test/test_jit_cuda_fuser.py [hunk 3]
```diff
@@ -406,6 +408,109 @@ def test_random_topo(self):
...
+
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING and GRAPH_EXECUTOR !=
+                     ProfilingMode.LEGACY, "Requires fusion optimization pass to be effective")
+    @skipIfRocm
+    def test_reduction(self):
+        for x in ([7, 8, 12], [12, 8, 7, 9, 15], [128, 16, 8, 32]):
+            # note that num_dim is exclusive from len(x), so we are not reducing
+            # to single element (codegen limitation at this moment)
+            for num_reduce_dim in range(1, len(x)):
```
Comment (csarofeen) [inline test/test_jit_cuda_fuser.py:459]: I don't think we'll support full reductions for a little while, may want to make this `len(x)-1`

File: test/test_jit_cuda_fuser.py [hunk 4]
```diff
@@ -406,6 +408,109 @@ def test_random_topo(self):
...
+        dtype = torch.float
+        device = "cuda"
+        x = torch.randn(sizes, dtype=dtype, device=device)
+        y = torch.randn(sizes, dtype=dtype, device=device)
+        z = torch.randn(sizes, dtype=dtype, device=device)
+
+        def t(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
+            o = torch.add(x, y)
+            o = torch.sum(o, dim=[0])
+            o = torch.add(o, z)
```
Comment (csarofeen) [inline test/test_jit_cuda_fuser.py:478]: We aren't going to support pwise fusion on outputs in the first iteration, do we still want to include this?

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: if we need to start scrubbing everything in codegenerator from int -> int64 it will take some real doing.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 5]
```diff
@@ -31,11 +31,63 @@ typedef Expr* CgOp;
...
+  std::unordered_map<int, int> coalesc_permute;
+  for (size_t i = 0; i < reduction_axes.size(); i++) {
+    size_t new_pos = i + n_dims - reduction_axes.size();
+    if (new_pos == reduction_axes[i]) {
+      break;
+    } else {
+      coalesc_permute[reduction_axes[i]] = new_pos;
+    }
+  }
+  tv->reorder(coalesc_permute);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Maybe need
```
if(!coalesc_permute.empty())
```
otherwise, maybe leave identity transforms in the reorder map?

File: torch/csrc/jit/codegen/cuda/kernel.cpp [hunk 6]
```diff
@@ -541,8 +543,30 @@ void runKernel(
   // Naive launch config;
   size_t numel = outputs[0].numel();
 
-  // TODO: we can't randomly clap down this until we got striding.
-  const auto nBlocks = ceilDiv(numel, 128 * entry->unroll_factor_);
+  int blocks;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: please explicitly init all the locals, ex. int blocks = 0;

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 7]
```diff
@@ -102,46 +186,129 @@ class IrParser {
...
-        out->merge(0, 1);
-      // Split into 128 which will be bockDim.x
-      out->split(0, nthreads);
-      // Split by another 4 which will be our unroll factor
-      auto ur_factor = disable_unroll ? 1 : unroll_factor;
-      if (!disable_unroll) {
-        out->split(0, ur_factor);
-        cuda_kernel_->unroll_factor_ = ur_factor;
+      // TODO: has_reduction for scheudling should be done on a per output
+      //       tensor basis.
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:189]: This can be done pretty easily in the code generator. If you want to know whether or not a reduction has occurred in the history of an output. Please open an issue.

File: torch/csrc/jit/codegen/cuda/kernel.cpp [hunk 8]
```diff
@@ -541,8 +543,30 @@ void runKernel(
   // Naive launch config;
   size_t numel = outputs[0].numel();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: this can be const too, right?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 9]
```diff
@@ -102,46 +186,129 @@ class IrParser {
...
+      if (has_reduction) {
+        // TODO: this scheduling only works for a single reduction operation in
+        //       the fusion, in this case we can coalesc all reduction axes and
+        //       merge them together. (same applies to iteration axes)
+        // TODO: does this work for multiple outputs?
+
+        // query if fastest changing dimension (FCD) is a reduction
+        fcd_reduction = out->axis((int)out->nDims() - 1)->isReduction();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:197]: Doesn't this depend on the sorted stride order?

File: torch/csrc/jit/codegen/cuda/manager.cpp [hunk 10]
```diff
@@ -205,9 +212,31 @@ void runCudaFusionGroup(const Node* const fusion_node, Stack& stack) {
 
       auto tensor = at::empty_strided(sizes, strides, options);
       outputs.push_back(tensor);
+
+      // TODO: unsafe broadcast assumption;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/manager.cpp]: please elaborate, what exactly is the unsafe assuption?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 11]
```diff
@@ -102,46 +186,129 @@ class IrParser {
...
+          out->merge(0, 1);
+        }
+        // Merge all reduction dimensions
+        while (out->nDims() > 2) {
+          out->merge(1, 2);
+        }
+
+      } else {
+        // Merge all dimensions because we're only supporting pointwise
+        while (out->nDims() > 1)
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:218]: Do we pre-coalesce dimensions based on contiguity? I know I should remember if we do or don't. This will likely be critical for perf generally. We can merge dims for scheduling, but we shouldn't be merging contiguous dims unless they can't be considered contiguous because of the operations in the fusion.

File: torch/csrc/jit/codegen/cuda/parser.h [hunk 12]
```diff
@@ -26,6 +26,15 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
+#define PW_THREAD_X 128
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/parser.h]: constexpr int PW_THREAD_X = 128; ...

NOTE: we should probably follow suit with the rest of PyTorch C++ code and use the kFoo naming for compile time constants (ex. kPwThreadX instead of PW_THREAD_X). I'll leave it up to you if you want to make these changes, or we can do a mass rename pass later.

File: torch/csrc/jit/codegen/cuda/parser.h [hunk 13]
```diff
@@ -26,6 +26,15 @@ namespace jit {
...
+#define FCD_REDUCTION_THREAD_X 128
+#define NON_FCD_REDUCTION_THREAD_X 32
+#define NON_FCD_REDUCTION_THREAD_Y 32
+
+TORCH_CUDA_API bool hasReductionNode(const Block* const block);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/parser.h]: please remove the 2nd const (const Block* block)

File: torch/csrc/jit/codegen/cuda/parser.h [hunk 14]
```diff
@@ -26,6 +26,15 @@ namespace jit {
...
+
+TORCH_CUDA_API bool isReductionNode(const Node* const node);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/parser.h]: ditto

File: torch/csrc/jit/codegen/cuda/partition.cpp [hunk 15]
```diff
@@ -57,6 +57,20 @@ inline bool isFusableNode(const Node* const node) {
...
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
 } // namespace
 
 bool isFusableCudaFusionGroup(const Node* const node) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/partition.cpp]: sigh, I see the extra consts are everywhere. I can do a separate pass to clean this up then...

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp [hunk 16]
```diff
@@ -141,8 +142,21 @@ class NaiveShapeTypePropagator {
         node->output()->setType(promoted_type);
         break;
       }
+      case aten::sum: {
+        auto out_type = node->input(0)->type()->cast<TensorType>();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/shape_inference.cpp]: these can all be 'const auto'

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp [hunk 17]
```diff
@@ -154,6 +168,27 @@ class NaiveShapeTypePropagator {
   }
 
  protected:
+  TensorTypePtr unary_reduce_type(
+      TensorTypePtr const& op,
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/shape_inference.cpp]: nit: const TensorTypePtr& ?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 18]
```diff
@@ -102,46 +186,129 @@ class IrParser {
...
-      // Split into 128 which will be bockDim.x
-      out->split(0, nthreads);
-      // Split by another 4 which will be our unroll factor
-      auto ur_factor = disable_unroll ? 1 : unroll_factor;
-      if (!disable_unroll) {
-        out->split(0, ur_factor);
-        cuda_kernel_->unroll_factor_ = ur_factor;
+      // TODO: has_reduction for scheudling should be done on a per output
+      //       tensor basis.
+      if (has_reduction) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:190]: Can you split this out at a higher level? something like SchedulePWFusion and ScheduleReductionFusion?

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp [hunk 19]
```diff
@@ -154,6 +168,27 @@ class NaiveShapeTypePropagator {
...
+        op->scalarType().has_value() && op->device().has_value() &&
+            op->sizes().isComplete(),
+        "requires complete shape on input");
+    std::vector<int64_t> output_size;
+    std::vector<int64_t> input_size = *op->sizes().concrete_sizes();
+    for (size_t i = 0; i < input_size.size(); i++) {
+      if (std::find(dims.begin(), dims.end(), i) == dims.end()) {
+        output_size.emplace_back(input_size[i]);
+      } else if (keepdim) {
+        output_size.emplace_back(1);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/shape_inference.cpp:187]: what is the meaning of 1 here?

**Review by tlemo — APPROVED** [review]
Looks good to me. I left a few small notes and questions.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 20]
```diff
@@ -102,46 +186,129 @@ class IrParser {
...
+        // scheduling of inputs shouldn't change with different fcd_reduction
+        intermediate->computeAt(out_tv, -2);
+        if (fcd_reduction) {
+          out_tv->axis(0)->parallelize(ParallelType::BIDx);
+        } else {
+          out_tv->axis(0)->parallelize(ParallelType::BIDx);
+          out_tv->axis(1)->parallelize(ParallelType::TIDx);
+        }
+      }
+      // Run through intermediates, unroll, and bind their axes
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Runs through all values, not just intermediates.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 21]
```diff
@@ -457,14 +646,57 @@ class IrParser {
...
+    }
   }
 
   void processJitNode(const JitOp* node) {
     if (node->kind() == prim::Constant) {
       // partition doesn't take constant node explicitly, but it does and copy
       // constant into subgraph. So we need to register constants in codegen IR;
       for (auto output : node->outputs()) {
-        TORCH_CHECK(registerScalar(output));
+        TORCH_CHECK(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Should this be internal_assert instead of torch check?

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp (hunk 19 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/shape_inference.cpp:187]: for `x = torch.randn(10, 10).sum(axis=0, keep_dim=True)`, `x.size()` would give us `[1, 10]`; if we do `keep_dim=False`, `x.size()` gives `[10]`.

So when keepdim is true, we want to keep the dimension by setting its size to 1.

File: test/test_jit_cuda_fuser.py (hunk 3 above)
Comment (jjsjann123) [reply test/test_jit_cuda_fuser.py:459]: range is exclusive for the second argument, so the number we are getting are [1, len(x) ), which is basically `1, 2, ..., len(x) - 1 `

File: test/test_jit_cuda_fuser.py (hunk 4 above)
Comment (jjsjann123) [reply test/test_jit_cuda_fuser.py:478]: No we are not supporting that.

This test is to make sure that graph partition does it properly and would *NOT* take the second `add` in the graph and give codegen error.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 5 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Good catch!

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 9 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:197]: That's true. I'm not handling sorted stride / contiguity in current integration yet.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 11 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:218]: Same argument as above, we should but we are not there yet.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 18 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:190]: Can we do that after the next refactor (2 PRs later)? I want to isolate scheduling logic to a `magic scheduler` interface and there I can have things separate out better.

File: test/test_jit_cuda_fuser.py (hunk 3 above)
Comment (csarofeen) [reply test/test_jit_cuda_fuser.py:459]: 🤦‍♂️

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: reverted.

File: torch/csrc/jit/codegen/cuda/parser.h (hunk 12 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.h]: Updated. Feel free to yell at me if you notice I missed anything :)

File: torch/csrc/jit/codegen/cuda/partition.cpp (hunk 15 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/partition.cpp]: I updated all of `const T* const` in integration code, but left the codegen part alone. :stuck_out_tongue: 
I did notice our migration to remove the const pointer. Out of curiosity, what's to hate const pointer?
